src/controllers/version/downloads.rs

```diff
@@ -19,59 +21,82 @@ pub fn download(req: &mut dyn RequestExt) -> EndpointResult {
     let version = req.params()["version"].as_str();
 
     let mut log_metadata = None;
-    match req.db_conn() {
-        Ok(conn) => {
-            use self::versions::dsl::*;
-
-            // Returns the crate name as stored in the database, or an error if we could
-            // not load the version ID from the database.
-            let (version_id, canonical_crate_name) = app
-                .instance_metrics
-                .downloads_select_query_execution_time
-                .observe_closure_duration(|| {
-                    versions
-                        .inner_join(crates::table)
-                        .select((id, crates::name))
-                        .filter(Crate::with_name(&crate_name))
-                        .filter(num.eq(version))
-                        .first::<(i32, String)>(&*conn)
-                })?;
-
-            if canonical_crate_name != crate_name {
-                app.instance_metrics
-                    .downloads_non_canonical_crate_name_total
-                    .inc();
-                log_metadata = Some(("bot", "dl"));
-            }
-            crate_name = canonical_crate_name;
+
+    match app
+        .version_id_cacher
+        .entry(format!("{}:{}", crate_name, version))
```

Comment:
`DashMap::entry` acquires a *write* lock, even if the entry is present in the map. This would cause every request to do a write lock on the assigned shard, which we should try to avoid. We should do something similar to [DownloadsCounter](https://github.com/rust-lang/crates.io/blob/4513c507c60bb8ec03fc5d859ceff56264cb2a94/src/downloads_counter.rs#L48-L65), where we first call `.get` to try and use the read lock and only if it fails call `.entry`.

Reply:
Excellent point, I'll switch to `.get`.